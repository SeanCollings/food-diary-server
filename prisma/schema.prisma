// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
