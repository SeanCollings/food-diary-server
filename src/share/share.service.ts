import { formatSummaryData } from '@/utils/modules/summary-utils';
import { PrismaService } from '@/prisma.service';
import { createGuid } from '@/utils/string-utils';
import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { getSortedDatesInRange } from '@/utils/date-utils';

@Injectable()
export class ShareService {
  constructor(private prisma: PrismaService) {}

  async getSharedSummary(link: string, dateFrom: string, dateTo: string) {
    const datesInRange = getSortedDatesInRange(dateFrom, dateTo);

    const shareLink = await this.prisma.shareLink.findFirst({
      where: { link, AND: { isShared: true } },
      include: { user: { select: { name: true } } },
    });

    if (!shareLink) {
      throw new Error(`Share page with link: ${link} does not exist`);
    }

    const diaryDays = await this.prisma.diaryDay.findMany({
      where: {
        userId: shareLink.userId,
        date: { in: datesInRange },
      },
    });

    const shareData = formatSummaryData(datesInRange, diaryDays);

    return {
      user: shareLink.user.name,
      dates: datesInRange,
      totalDays: datesInRange.length,
      data: { ...shareData },
    };
  }

  async generateShareLink(userId: number) {
    const shareLink = createGuid();

    try {
      await this.prisma.shareLink.upsert({
        where: { userId },
        create: {
          userId,
          link: shareLink,
          isShared: true,
        },
        update: {
          link: shareLink,
        },
      });

      return { shareLink };
    } catch (e) {
      if (e instanceof Prisma.PrismaClientKnownRequestError) {
        if (e.code === 'P2002') {
          throw new Error(
            'There is a unique constraint violation, a new ShareLink cannot be created with this userid',
          );
        }
      }
      throw e;
    }
  }

  async linkShareable(userId: number, isShared: boolean) {
    return this.prisma.shareLink.update({
      where: { userId },
      data: { isShared },
    });
  }
}
