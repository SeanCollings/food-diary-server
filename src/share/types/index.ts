import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import {
  IsDateString,
  IsNotEmpty,
  IsString,
  IsUUID,
  MaxLength,
  MinLength,
} from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetSharedSummaryQuery {
  @IsNotEmpty()
  @MinLength(36)
  @MaxLength(36)
  @IsString()
  @IsUUID()
  link: string;

  @IsNotEmpty()
  @IsDateString()
  dateFrom: string;

  @IsNotEmpty()
  @IsDateString()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
