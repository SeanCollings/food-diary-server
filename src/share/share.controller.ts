import { JwtAuthGuard } from '@/auth/jwt-auth.guard';
import { DEFAULT_ERROR_MSG } from '@/lib/validation/validation.constants';
import {
  Controller,
  Put,
  UseGuards,
  Request,
  InternalServerErrorException,
  Body,
  Get,
  Query,
} from '@nestjs/common';
import { LinkShareableDto } from './dtos/link-shareable.dto';
import { ShareService } from './share.service';
import { GetSharedSummaryQuery, RequestWithUser } from './types';

@Controller('share')
export class ShareController {
  constructor(private shareService: ShareService) {}

  @Get('/')
  async getSharedSummary(@Query() query: GetSharedSummaryQuery) {
    try {
      return await this.shareService.getSharedSummary(
        query.link,
        query.dateFrom,
        query.dateTo,
      );
    } catch (err) {
      console.error('[share::_get_]:', err.message);
      throw new InternalServerErrorException(err.message || DEFAULT_ERROR_MSG);
    }
  }

  @Put('/generate-link')
  @UseGuards(JwtAuthGuard)
  async generateShareLink(@Request() req: RequestWithUser) {
    try {
      return await this.shareService.generateShareLink(req.user.userId);
    } catch (err) {
      console.error('[share::_put_generate-link]:', err.message);
      throw new InternalServerErrorException(err.message || DEFAULT_ERROR_MSG);
    }
  }

  @Put('/link-shareable')
  @UseGuards(JwtAuthGuard)
  async linkShareable(
    @Request() req: RequestWithUser,
    @Body() body: LinkShareableDto,
  ) {
    try {
      return await this.shareService.linkShareable(
        req.user.userId,
        body.isShared,
      );
    } catch (err) {
      console.error('[share::_put_link-shareable]:', err.message);
      throw new InternalServerErrorException(err.message || DEFAULT_ERROR_MSG);
    }
  }
}
