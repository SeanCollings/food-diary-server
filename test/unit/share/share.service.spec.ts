import { Test, TestingModule } from '@nestjs/testing';
import { ShareService } from '@/share/share.service';
import { PrismaService } from '@/prisma.service';
import { mockDeep, DeepMockProxy } from 'jest-mock-extended';
import { Prisma, PrismaClient } from '@prisma/client';

jest.mock('@/utils/string-utils', () => ({
  createGuid: jest.fn().mockReturnValue('mock_guid'),
}));

describe('ShareService', () => {
  let service: ShareService;
  let prisma: DeepMockProxy<PrismaClient>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ShareService, PrismaService],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .compile();

    service = module.get(ShareService);
    prisma = module.get(PrismaService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getSharedSummary', () => {
    it('should return a shared summary', async () => {
      prisma.shareLink.findFirst.mockResolvedValue({
        id: 1,
        isShared: true,
        link: 'share_link',
        userId: 1234,
        user: { name: 'Test User' },
      } as any);
      prisma.diaryDay.findMany.mockResolvedValue([]);

      const result = await service.getSharedSummary(
        'mock_link',
        '2023-04-02',
        '2023-04-13',
      );

      expect(result).toMatchSnapshot();
    });

    it('should throw error if no shareLink exists', async () => {
      prisma.shareLink.findFirst.mockResolvedValue(undefined as any);

      try {
        await service.getSharedSummary('mock_link', '2023-04-02', '2023-04-13');
      } catch (err) {
        expect(err.message).toMatchInlineSnapshot(
          `"Share page with link: mock_link does not exist"`,
        );
      }
    });
  });

  describe('generateShareLink', () => {
    it('should create and return a share-link', async () => {
      prisma.shareLink.upsert.mockResolvedValue({} as any);

      const result = await service.generateShareLink(1234);

      expect(result).toMatchInlineSnapshot(`
        {
          "shareLink": "mock_guid",
        }
      `);
    });

    it('should handle errors', async () => {
      prisma.shareLink.upsert.mockRejectedValue({ message: 'Error occurred' });

      try {
        await service.generateShareLink(1234);
      } catch (err) {
        expect(err.message).toMatchInlineSnapshot(`"Error occurred"`);
      }
    });

    it('should handle Prisma.PrismaClientKnownRequestErrors with code P2002', async () => {
      const error = new Prisma.PrismaClientKnownRequestError(
        'Error occurred',
        'P2002',
        'v1',
      );
      prisma.shareLink.upsert.mockRejectedValue(error);

      try {
        await service.generateShareLink(1234);
      } catch (err) {
        expect(err.message).toMatchInlineSnapshot(
          `"There is a unique constraint violation, a new ShareLink cannot be created with this userid"`,
        );
      }
    });

    it('should handle Prisma.PrismaClientKnownRequestErrors that is not code P2002', async () => {
      const error = new Prisma.PrismaClientKnownRequestError(
        'Error occurred',
        'Other',
        'v1',
      );
      prisma.shareLink.upsert.mockRejectedValue(error);

      try {
        await service.generateShareLink(1234);
      } catch (err) {
        expect(err.message).toMatchInlineSnapshot(`"Error occurred"`);
      }
    });
  });

  describe('linkShareable', () => {
    it('should update a sharelink shareability and return that shareLink', async () => {
      prisma.shareLink.update.mockResolvedValue({
        id: 555,
        userId: 1234,
        isShared: true,
        link: 'mock_link',
      });

      const result = await service.linkShareable(1234, true);
      expect(result).toMatchInlineSnapshot(`
        {
          "id": 555,
          "isShared": true,
          "link": "mock_link",
          "userId": 1234,
        }
      `);
    });
  });
});
