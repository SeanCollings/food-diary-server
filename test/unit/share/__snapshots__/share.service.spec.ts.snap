// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareService getSharedSummary should return a shared summary 1`] = `
{
  "data": {},
  "dates": [
    "2023-04-01T00:00:00.000Z",
    "2023-04-02T00:00:00.000Z",
    "2023-04-03T00:00:00.000Z",
    "2023-04-04T00:00:00.000Z",
    "2023-04-05T00:00:00.000Z",
    "2023-04-06T00:00:00.000Z",
    "2023-04-07T00:00:00.000Z",
    "2023-04-08T00:00:00.000Z",
    "2023-04-09T00:00:00.000Z",
    "2023-04-10T00:00:00.000Z",
    "2023-04-11T00:00:00.000Z",
    "2023-04-12T00:00:00.000Z",
  ],
  "totalDays": 12,
  "user": "Test User",
}
`;
